README.rst
===================
Text-based-switcher
===================

Text-based window switcher for Unity

The visual switcher is very poor at providing context. E.g. the browser window
thumbnails are too small to tell apart. 

This switcher shows a list of window titles and allows you to choose one
by typing first character(s) of the sought window. Press return to bring the
window to the front.

All credit goes to the original author, Jacob Vlijm. Thanks!
See http://askubuntu.com/a/648800/20835

Usage
=====

The script needs ``wmctrl`` and ``xprop``. On Ubuntu:

.. code:: console

    $ sudo apt-get install wmctrl x11-utils

Install from PyPI:

.. code:: console

    $ sudo pip install text_based_switcher

Test-run it:

.. code:: console

    $ list_windows --workspace
    $ list_windows --application
    $ list_windows --window

If it works, add one or more of the preferred commands to one or more shortcut
keys: choose:
_System Settings_ > _Keyboard_ > _Shortcuts_ > _Custom Shortcuts_.

Click the "+" and add the command.

If you do a local install (``--user`` option), Compiz won't get the right
Python path.
You can fix this by hacking ``sys.path`` in the installed script.

TODO
====

- Add wmctrl or ruamel.emwh as dependency
- Add xprop as dependency
- Don't call commands, use Python libraries
- Write some tests


Developer notes
===============

Please use a virtualenv to maintain this package, but I should not need to say that.

Grab the source from the SCM repository:

.. code:: console

  $ python setup.py develop
  $ pip install text-based-switcher[dev]

If you install 

Run the tests (there are no tests yet!):

.. code:: console

  $ python setup.py test
  $ python run_tests.py


Links
=====

Project home page

  https://github.com/jean/text-based-switcher

Source code

  https://github.com/jean/text-based-switcher

Issue tracker

  https://github.com/jean/text-based-switcher/issues


CHANGES.rst
Changes log
===========

0.0.1
-----

- First version by Jacob Vlijm as an answer on AskUbuntu:
  http://askubuntu.com/a/648800/20835

0.0.2
-----

- Package creation with ``bobtemplates.gillux``
  [jean]


tests/__init__.py
# -*- coding: utf-8 -*-
"""\
===================
text_based_switcher
===================

Tests package
"""
import sys
if sys.version_info < (2, 7):
    import unittest2 as unittest
else:
    import unittest

from .resources import tests_directory

def all_tests():
    return unittest.defaultTestLoader.discover(tests_directory)


tests/test_package.py
# -*- coding: utf-8 -*-
"""Unit tests of text_based_switcher"""

import sys
nottest = istest = lambda obj: obj  # Fake decoratos provided by nose
if sys.version_info < (2, 7):
    import unittest2 as unittest
else:
    import unittest

import resources


###
# Test layers
###

@nottest
def setUpModule():
    # Setups for all tests of this module
    return

@nottest
def tearDownModule():
    # Cleanups for all tests of this module
    return

###
# Test cases
###

class SomeTest(unittest.TestCase):
    """FIXME: describe here what you are testing
    """
    @classmethod
    def setUpClass(cls):
        # Setups before first test of this class
        return

    @classmethod
    def tearDownClass(cls):
        # Cleanups after last test of this class
        return

    def setUp(self):
        # Setups before each test of this class
        return

    def tearDown(self):
        # Cleanups after each test of this class
        return

    def test_fake(self):
        """FIXME: provide your first test here
        """
        self.assertEqual(1, 1)
        return


tests/test_doctests.py
# -*- coding: utf-8 -*-
"""Doctests of text_based_switcher"""

import doctest
import glob

from resources import tests_abs_path

# Add resources available in the file doctests
filedoctest_globs = {
    'dummy': lambda: None
}

def filedoctest_setup(test):
    # Add fixtures for your doctest files
    # See https://docs.python.org/2.7/library/doctest.html#doctest.DocFileSuite
    return

def filedoctest_teardown(test):
    # Cleanup fixtures for your doctest files
    # See https://docs.python.org/2.7/library/doctest.html#doctest.DocFileSuite
    return

def load_tests(loader, tests, ignore):
    # See https://docs.python.org/2.7/library/doctest.html#unittest-api

    # Run the tests in the .rst files of this directory
    doctest_files = glob.glob(tests_abs_path('test_*.rst'))
    tests.addTests(doctest.DocFileSuite(*doctest_files, module_relative=False,
                                        setUp=filedoctest_setup, tearDown=filedoctest_teardown,
                                        globs=filedoctest_globs))

    # Run the doctests in the various modules
    import text_based_switcher
    modules_with_doctests = (text_based_switcher,)
    for module in modules_with_doctests:
        tests.addTests(doctest.DocTestSuite(module))
    return tests


src/text_based_switcher/__init__.py
# -*- coding: utf-8 -*-
"""
===================
text_based_switcher
===================

Text-based window switcher for Unity
"""

import sys
import logging
import pkg_resources

# Custom logger
LOG = logging.getLogger(name=__name__)
if sys.version_info < (2, 7):
    class NullHandler(logging.Handler):
        """Copied from Python 2.7 to avoid getting `No handlers could be found
        for logger "xxx"` http://bugs.python.org/issue16539
        """
        def handle(self, record):
            pass
        def emit(self, record):
            pass
        def createLock(self):
            self.lock = None
else:
    from logging import NullHandler

LOG.addHandler(NullHandler())

# PEP 396 style version marker
try:
    __version__ = pkg_resources.get_distribution(u'text-based-switcher').version
except:
    LOG.warning("Could not get the package version from pkg_resources")
    __version__ = 'unknown'


src/text_based_switcher/list_windows.py
#!/usr/bin/env python3

# Originally written by Jacob Vlijm, see http://askubuntu.com/a/648800/20835

import subprocess
import socket
import sys

from optparse import OptionParser

parser = OptionParser()
parser.add_option("-a", "--application",
        action="store_true",
        help="Sort on application name")
parser.add_option("-w", "--workspace",
        action="store_true",
        help="Sort on workspace number")
parser.add_option("-t", "--window",
        action="store_true",
        help="Sort on window title")

(options, __) = parser.parse_args()

# just a helper function
get = lambda cmd: subprocess.check_output([
    "/bin/bash", "-c", cmd
    ]).decode("utf-8")


def normal_window(w_id):
    """ function to distinguish "normal" windows from other types 
    
    (like the desktop etc)
    """
    w_type = get("xprop -id "+w_id)
    if " _NET_WM_WINDOW_TYPE_NORMAL" in w_type:
        return True
    else:
        return False


# analyse viewport data, to be able to calculate relative/absolute position of
# windows and current viewport.
xrandr_out = get("xrandr").split()
i = xrandr_out.index("current")
resolution = [
        int(xrandr_out[i+1]),
        int(xrandr_out[i+3].replace(",", ""))]

wmctrl_out = get("wmctrl -d").split()
virtual_resolution = [int(n) for n in wmctrl_out[3].split("x")]
ws_columns = int(virtual_resolution[0] / resolution[0])
ws_rows = int(virtual_resolution[1] / resolution[1])

vector = [int(n) for n in wmctrl_out[5].split(",")]
current_viewport = int(
        (vector[1]/resolution[1]) * ws_columns + 
        (vector[0]/resolution[0])+1)

# split windowdata by machine name
machine_name = socket.gethostname()
window_list = [
        [l.strip() for l in w.split(machine_name)]
        for w in get("wmctrl -lpG").splitlines()]

# split first section of window data
for i, w in enumerate(window_list):
    window_list[i][0] = window_list[i][0].split()

# filter only "real" windows
window_list = [w for w in window_list if normal_window(w[0][0])]

# adding the viewport to the window's data
for window in window_list:
    window.append(get("ps -p "+window[0][2]+" -o comm=").strip())
    relative_location = [int(n) for n in window[0][3:5]]
    absolute_location = [
            relative_location[0]+vector[0],
            relative_location[1]+vector[1]]
    viewport = int(
            (absolute_location[1]/resolution[1]) * 
            ws_columns +
            (absolute_location[0]/resolution[0])+1)

    if viewport == current_viewport:
        viewport = "*"+str(viewport)
    else:
        viewport = str(viewport)
    if viewport.startswith('-'):
        viewport = viewport.replace('-', '_')
    window.append(viewport)

# set sorting rules
if options.application:
    window_list.sort(key=lambda x: (x[-2].lower(), x[-1], x[-3].lower()))
elif options.workspace:
    window_list.sort(key=lambda x: (x[-1], x[-3].lower(), x[-2].lower()))
elif options.window:
    window_list.sort(key=lambda x: (x[-3].lower(), x[-1], x[-2].lower()))

# calculate width and height of the zenity window:
# height = 140px + 23px per line
height = str(140+(len(window_list)*23))

# width = 250px + 8px per character (of the longest window title)
width = str(250+(max([len(window[-3]) for window in window_list])*8))

escape = lambda s: s.replace('"', '\\"')

# define the zenity window's content
cmd = (u"zenity --list --hide-column=4 --print-column=4 "
        "--title='Window list' "
        "--width=%s "% width +
        "--height=%s "% height +
        "--column='Window name' "
        "--column='Application name' "
        "--column='Workspace' "
        "--column='w_id' " + " ".join(
            [(" ").join(
                ['"%s"'% escape(window[-3]), # window title
                 '"%s"'% escape(window[-2]), # application name
                 '"%s"'% escape(window[-1]), # workspace name
                 '"%s"'% escape(window[0][0])]
                 ) for window in window_list]))

# finally, call the window list
try:
    w_id = subprocess.check_output(["/bin/bash", "-c", cmd]
        ).decode("utf-8").split("|")[0]
    subprocess.Popen(["wmctrl", "-ia", w_id])
except subprocess.CalledProcessError:
    pass


src/text_based_switcher/__main__.py
# -*- coding: utf-8 -*-
"""
===================
text_based_switcher
===================

Commandline handling
"""
# Originally written by Jacob Vlijm, see http://askubuntu.com/a/648800/20835

import logging
import optparse
import socket
import subprocess
import sys

from text_based_switcher import __version__, LOG


def main(argv=sys.argv):
    """Main function called from console command
    """
    logging.basicConfig()
    exit_code = 1
    try:
        app = Application(argv)
        app.run()
        exit_code = 0
    except KeyboardInterrupt:
        exit_code = 0
    except Exception as exc:
        LOG.exception(exc)
    sys.exit(exit_code)


class Application(object):
    """The main Application class

    :param argv: The command line as a list as ``sys.argv``
    """
    def __init__(self, argv):
        # ap = argparse.ArgumentParser()
        # ap.add_argument('--version', action='version', version=__version__)
        # self.args = ap.parse_args(args=argv[1:])
        parser = optparse.OptionParser()
        parser.add_option("-a", "--application",
                action="store_false",
                help="Sort on application name")
        parser.add_option("-w", "--workspace",
                action="store_true",
                help="Sort on workspace number")
        parser.add_option("-t", "--window",
                action="store_false",
                help="Sort on window title")

        (options, __) = parser.parse_args()
        self.options = options

    def get_output(self, cmd):
        return subprocess.check_output(["/bin/bash", "-c", cmd]).decode("utf-8")

    def normal_window(self, w_id):
        """ function to distinguish "normal" windows from other types 
        
        (like the desktop etc)
        """
        w_type = self.get_output("xprop -id "+w_id)
        if " _NET_WM_WINDOW_TYPE_NORMAL" in w_type:
            return True
        else:
            return False

    def run(self):
        # analyse viewport data, to be able to calculate relative/absolute position of
        # windows and current viewport.
        xrandr_out = self.get_output("xrandr").split()
        i = xrandr_out.index("current")
        resolution = [
                int(xrandr_out[i+1]),
                int(xrandr_out[i+3].replace(",", ""))]

        wmctrl_out = self.get_output("wmctrl -d").split()
        virtual_resolution = [int(n) for n in wmctrl_out[3].split("x")]
        ws_columns = int(virtual_resolution[0] / resolution[0])
        ws_rows = int(virtual_resolution[1] / resolution[1])

        vector = [int(n) for n in wmctrl_out[5].split(",")]
        current_viewport = int(
                (vector[1]/resolution[1])*ws_columns + 
                (vector[0]/resolution[0])+1)

        # split windowdata by machine name
        machine_name = socket.gethostname()
        window_list = [
                [l.strip() for l in w.split(machine_name)]
                for w in self.get_output("wmctrl -lpG").splitlines()]

        # split first section of window data
        for i, w in enumerate(window_list):
            window_list[i][0] = window_list[i][0].split()

        # filter only "real" windows
        window_list = [w for w in window_list if self.normal_window(w[0][0])]

        # adding the viewport to the window's data
        for window in window_list:
            window.append(
                    self.get_output("ps -p "+window[0][2]+" -o comm=").strip())
            relative_location = [int(n) for n in window[0][3:5]]
            absolute_location = [
                    relative_location[0]+vector[0],
                    relative_location[1]+vector[1]]
            viewport = int(
                    (absolute_location[1]/resolution[1])*ws_columns +
                    (absolute_location[0]/resolution[0])+1)

            if viewport == current_viewport:
                viewport = "*"+str(viewport)
            else:
                viewport = str(viewport)
            if viewport.startswith('-'):
                viewport = viewport.replace('-', '_')
            window.append(viewport)

        # set sorting rules
        if self.options.application:
            window_list.sort(
                    key=lambda x: (x[-2].lower(), x[-1], x[-3].lower()))
        elif self.options.workspace:
            window_list.sort(
                    key=lambda x: (x[-1], x[-3].lower(), x[-2].lower()))
        elif self.options.window:
            window_list.sort(
                    key=lambda x: (x[-3].lower(), x[-1], x[-2].lower()))

        # calculate width and height of the zenity window:
        # height = 140px + 23px per line
        height = str(140+(len(window_list)*23))

        # width = 250px + 8px per character (of the longest window title)
        width = str(250+(max([len(window[-3]) for window in window_list])*8))

        escape = lambda s: s.replace('"', '\\"')

        # define the zenity window's content
        cmd = (u"zenity --list --hide-column=4 --print-column=4 "
                "--title='Window list' "
                "--width=%s "% width +
                "--height=%s "% height +
                "--column='Window name' "
                "--column='Application name' "
                "--column='Workspace' "
                "--column='w_id' " + " ".join(
                    [(" ").join(
                        ['"%s"'% escape(window[-3]), # window title
                         '"%s"'% escape(window[-2]), # application name
                         '"%s"'% escape(window[-1]), # workspace name
                         '"%s"'% escape(window[0][0])]
                         ) for window in window_list]))

        # finally, call the window list
        try:
            w_id = subprocess.check_output(["/bin/bash", "-c", cmd]
                ).decode("utf-8").split("|")[0]
            subprocess.Popen(["wmctrl", "-ia", w_id])
        except subprocess.CalledProcessError:
            pass




if __name__ == '__main__':
    main()
